### PR TITLE
spec loading cleanup

### PR DESCRIPTION
**Pull request by debris**
the goal of this pr is to cleanup loading spec and get rid of **all** unwraps (or at least as many as possible).

changes:
- in json spec files:
  - replaced `engineName` with `engine`
  - moved engine specific params to `engine` field
  - seal fields are grouped together
- completely removed usages of `FromJson` trait from `ethcore`
- removed `toEngine` method from spec. `engine` is now a spec property
- `spec` is no longer a property of `engine`
- engine params are no longer decoded from rlp.

This pr is so big mostly because I had to change a lot of test functions...

todo:
- update spec wiki

### CODE REVIEW

**Comment by NikVolf** [comment]
looks great

File: ethcore/src/json_tests/state.rs [hunk 1]
```diff
@@ -30,9 +30,9 @@ pub fn json_chain_test(json_data: &[u8], era: ChainEra) -> Vec<String> {
 	let tests = ethjson::state::Test::load(json_data).unwrap();
 	let mut failed = Vec::new();
 	let engine = match era {
-		ChainEra::Frontier => ethereum::new_mainnet_like(),
-		ChainEra::Homestead => ethereum::new_homestead_test(),
-	}.to_engine().unwrap();
+		ChainEra::Frontier => ethereum::new_mainnet_like().engine,
+		ChainEra::Homestead => ethereum::new_homestead_test().engine
```
Comment (gavofyork) [inline ethcore/src/json_tests/state.rs]: trailing comma missing.

File: ethcore/src/spec/seal.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Spec seal.
+
+use util::rlp::*;
+use util::hash::{H64, H256};
+use ethjson;
+
+/// Classic ethereum seal.
+pub struct Ethereum {
```
Comment (gavofyork) [inline ethcore/src/spec/seal.rs]: shouldn't really be here. any ethereum-specific stuff should be contained in the `ethereum` subdir.

File: ethcore/src/spec/spec.rs [hunk 3]
```diff
@@ -21,55 +21,51 @@ use engine::*;
...
-	}
+use ethjson;
+
+/// Parameters common to all engines.
+#[derive(Debug, PartialEq, Clone)]
+pub struct CommonParams {
+	/// Account start nonce.
+	pub account_start_nonce: U256,
+	/// Frontier compatibility mode limit.
+	pub frontier_compatibility_mode_limit: u64,
```
Comment (gavofyork) [inline ethcore/src/spec/spec.rs]: should really be an Ethash property.

File: ethcore/src/spec/spec.rs [hunk 4]
```diff
@@ -21,55 +21,51 @@ use engine::*;
...
 
 /// Parameters for a block chain; includes both those intrinsic to the design of the
 /// chain and those to be interpreted by the active chain engine.
-#[derive(Debug)]
 pub struct Spec {
 	/// User friendly spec name
 	pub name: String,
 	/// What engine are we using for this?
-	pub engine_name: String,
+	pub engine: Box<Engine>,
```
Comment (gavofyork) [inline ethcore/src/spec/spec.rs]: `Spec` is a _specification_ that describes the consensus engine. It was not designed to be used as a parent instance. Refactoring it in this way probably has unforseen consequences. `Engine` is meant to be the parent type to be passed around.

File: ethcore/src/spec/spec.rs [hunk 5]
```diff
@@ -101,18 +97,38 @@ pub struct Spec {
 	genesis_state: PodState,
 }
 
-#[cfg_attr(feature="dev", allow(wrong_self_convention))] // because to_engine(self) should be to_engine(&self)
-impl Spec {
-	/// Convert this object into a boxed Engine of the right underlying type.
-	// TODO avoid this hard-coded nastiness - use dynamic-linked plugin framework instead.
```
Comment (gavofyork) [inline ethcore/src/spec/spec.rs]: Please do not remove TODOs without doing them.

File: ethcore/src/spec/spec.rs [hunk 6]
```diff
@@ -101,18 +97,38 @@ pub struct Spec {
...
+impl From<ethjson::spec::Spec> for Spec {
+	fn from(s: ethjson::spec::Spec) -> Self {
+		let builtins = s.accounts.builtins().into_iter().map(|p| (p.0.into(), From::from(p.1))).collect();
+		let g = Genesis::from(s.genesis);
+		let seal: GenericSeal = g.seal.into();
+		let params = CommonParams::from(s.params);
+		Spec {
+			name: s.name.into(),
+			params: params.clone(),
+			engine: match s.engine {
```
Comment (gavofyork) [inline ethcore/src/spec/spec.rs]: this `match` should be left in its own function; it's currently a stub for a plugin loader and the code should reflect that.

**Comment by gavofyork** [comment]
this is a rather disstressing change.

the engine factory was a temporary place-holder for a dynamically-loaded plugin architecture; this change doesn't help achieve that goal.

it rearchitects the `Spec`/`Engine` ownership relationship, reducing `Engine` to a second-class utility type unable to be used without an owning `Spec`. at best, the name(s) need changing.

**Comment by gavofyork** [comment]
removing `unwrap`s is great and all, but what happens if the data of the json is in incorrect format? there's really only three possibilities:
- panic (no better than `unwrap`);
- bubble-up error on call stack (i don't see any `try!`s);
- silently give default value (probably just bad).

**Comment by debris** [comment]
I will try address all questions and concerns here:

> removing `unwraps` is great and all, but what happens if the data of the json is in incorrect format?

Currently there is no single place where whole json spec is validated. we have multiple unwraps in spec, builtin and verification. Besides that, there are `try` in [client constructor](https://github.com/ethcore/parity/blob/71dd9fb2dfff24633ac2ca5f6620057cbecca404/ethcore/src/client/client.rs#L143). In my pr, the whole json spec is validated deserialized in a single [place](https://github.com/ethcore/parity/pull/858/files#diff-8c15897d65990126e87e06b4feefef7dR235).

> `Spec` is a *specification* that describes the consensus engine. It was not designed to be used as a parent instance. 

This may be my misunderstanding... I see `spec` as a chain specification. And `engine` as a chain property.

> Refactoring it in this way probably has unforseen consequences. Engine is meant to be the parent type to be passed around.

I had the same concerns, but then I noticed that the only spec function that we use* (excluding `Engine`) is `ensure_db_good`. And it's used only just after `spec` is loaded.

> it rearchitects the `Spec`/`Engine` ownership relationship, reducing `Engine` to a second-class utility type unable to be used without an owning `Spec`. 

I think it's the other way around. On master branch, it's very difficult to create `Ethash` engine without loading json spec. Now it's trivial.

> at best, the name(s) need changing. at worst there are unforseen consequences of the design alteration (e.g. can we have an `Engine` in test code without needing a JSON description or using the placeholder `Engine` factory?) - keeping the design centred around `Engine` (which actually does the work) and not a JSON deserialisation type was an important design consideration. there is no clear addressing of the consequences of the design alteration.

Using JSON is completely optional and nothing relies on it now.

> the engine factory was a temporary place-holder for a dynamically-loaded plugin architecture; this change doesn't help achieve that goal.

I'm aware of that. And there is the same issue with genesis seal. That's true that this pr doesn't help achieve that goal, but imo it does not make it more difficult.

File: ethcore/src/spec/spec.rs (hunk 5 above)
Comment (debris) [inline ethcore/src/spec/spec.rs]: sure, my bad. Will bring it back

File: ethcore/src/spec/spec.rs (hunk 3 above)
Comment (debris) [inline ethcore/src/spec/spec.rs]: it's also used in `NullEngine` so I assumed that it's a property common to all engines

**Comment by debris** [comment]
I understand that this pr introduces two things that could be done separately (deserialization && object ownership refactoring), but it was just much easier to do this things together.

File: ethcore/src/spec/spec.rs (hunk 3 above)
Comment (gavofyork) [inline ethcore/src/spec/spec.rs]: not necessarily.

**Comment by gavofyork** [comment]
`Engine` certainly isn't a chain property. It's an auxiliary, used for much of the chain mechanics which may be defined (and thus constructed) by a portion of the chain `Spec`.

File: ethcore/src/json_tests/executive.rs [hunk 7]
```diff
@@ -27,33 +27,6 @@ use substate::*;
 use tests::helpers::*;
 use ethjson;
 
-struct TestEngineFrontier {
```
Comment (gavofyork) [inline ethcore/src/json_tests/executive.rs]: why removed?

**Comment by gavofyork** [comment]
probably good to go in, but i'd reconsider the naming in light of the architecture alteration.

File: ethcore/src/json_tests/executive.rs (hunk 7 above)
Comment (debris) [inline ethcore/src/json_tests/executive.rs]: Exactly the same structure was defined in `ethcore/src/tests/helpers.rs`

**Comment by debris** [comment]
Do you have any suggestions?

**Comment by gavofyork** [comment]
Having thought about it `Chain` or `ChainSpec` is the only thing I can think of which sounds as though it "owns" `Engine`. But I don't really like double-barrel names and just "Chain" doesn't seem much better than "Spec".

**Pull request merged**